wal/wal.go

```diff
@@ -408,15 +410,21 @@ func (w *FileWAL) Replay(firstIndex uint64, handler func(tx uint64, record *walp
 
 	level.Debug(w.logger).Log("msg", "replaying WAL", "first_index", firstIndex, "last_index", lastIndex)
 
+	tx := firstIndex
 	defer func() {
-		// recover the panic to print more context. Exit afterwards regardless.
-		if err := recover(); err != nil {
-			level.Error(w.logger).Log("msg", "replaying WAL failed", "path", w.path, "first_index", firstIndex, "last_index", lastIndex, "err", err)
-			panic(err)
+		// recover a panic of reading a transaction. Truncate the wal to the last valid transaction.
+		if r := recover(); r != nil {
+			level.Error(w.logger).Log("msg", "replaying WAL failed", "path", w.path, "first_index", firstIndex, "last_index", lastIndex, "offending_index", tx, "err", err)
+			if err = w.log.TruncateBack(tx - 1); err != nil {
```

Comment:
Hmm, I understand the reasoning, but it feels very aggressive to `TruncateBack`. Don't think you should change anything, just a note. Should we have a metric for this?

Reply:
The only thing we're missing out is the ability to manually copy out the WAL for debug purposes. We can track the number of times this occurs with the logging system.